### DISCUSSION TITLE
Correct way to asynchronously tryEnd() for streaming data

### DISCUSSION THREAD

**Discussion by cakepirate** [uNetworking/uWebSockets#1303] [Q&A]
Hi. I'm trying to rewrite the asyncfilestreamer example, with reference to the example in <https://github.com/uNetworking/uWebSockets.js/blob/master/examples/VideoStreamer.js>.

However, when I try to loop->defer each tryEnd() call for each chunk of data, and then loop->run() to block and complete execution before continuing onto the next chunk, the function does not exit the loop->defer call - am I misunderstanding the use of loop->defer?
```
template <bool SSL>
static void streamFile(uWS::HttpResponse<SSL> *res, AsyncFileReader *asyncFileReader)
{
    // asyncFileReader from https://github.com/uNetworking/uWebSockets/blob/master/examples/helpers/AsyncFileReader.h
    auto retStatus = std::make_pair<bool, bool>(false, false);
    while (!retStatus.second)
    {
        auto chunk = asyncFileReader->peek(res->getWriteOffset());
        asyncFileReader->loop->defer([res, asyncFileReader, chunk, &retStatus](){
            retStatus = res->tryEnd(chunk, asyncFileReader->getFileSize());
            if (!retStatus.first)
            {
                 // attach abort and onWritable handler
            }
        });
        asyncFileReader->loop->run();
    }
}
```

I also tried to loop->defer the entire while loop, but this results in non-asynchronous behaviour. What's the correct way to asynchronously perform this without resorting to the seemingly recursive pattern in the original asyncFile example?

**Comment by ghost** [comment]
Ugh, that code doesn't make any sense whatsoever.

**Reply by cakepirate** [reply]
Apologies, I edited it to make it hopefully more readable.
I'm confused about the correct use of loop defer. The documentation writes "Loop::defer takes a function (such as a lambda with data) and defers the execution of said function until the specified loop's thread is ready to execute the function in a single-threaded fashion on correct thread." - am I correct to interpret it as loop->defer allows one to send data in the response (via e.g. tryEnd) on the correct main thread, which allows one to start the streamFile function example above on a separate thread (so that the response is 'asynchronous'/ the server is freed to handle other requests)?
